ska_trend/centroid_dashboard/app.py

```diff
@@ -0,0 +1,1038 @@
+        action="store_true",
+    )
+    parser.add_argument(
+        "--remote-copy",
+        help="Copy asp L1 data from remote archive if not available locally",
+        action="store_true",
+    )
+    parser.add_argument(
+        "--skip-plots",
+        help="Skip generating plots (default=False)",
+        action="store_true",
+    )
+    parser.add_argument(
+        "--raise-exc",
+        help="Raise exceptions during processing for debugging (default=False)",
+        action="store_true",
+    )
+    parser.add_argument(
+        "--log-level", default="INFO", help="Logging level (default=INFO)"
+    )
+    return parser
+
+
+def raise_sporadic_exc_for_testing():
+    """Raise a quasi-sporadic exception for testing.
+
+    A handful of calls to this function are embedded in the code to raise an exception
+    sporadically for testing purposes. The exception is raised when the environment
+    variable `CENTROID_DASHBOARD_RAISE_EXC` is set to a positive integer.
+
+    For testing the exception handling, something like this::
+
+       env CENTROID_DASHBOARD_RAISE_EXC=11 python -m ska_trend.centroid_dashboard.app \
+          --start 2025:015 --stop 2025:017  --force
+    """
+    global SPORADIC_EXC_COUNT  # noqa: PLW0603
+    if not (n_ok := os.environ.get("CENTROID_DASHBOARD_RAISE_EXC")):
+        return
+
+    SPORADIC_EXC_COUNT += 1
+    if SPORADIC_EXC_COUNT % int(n_ok) == 0:
+        raise ValueError("Sporadic exception for testing")
+
+
+@functools.lru_cache()
+def get_index_template():
+    path = Path(__file__).parent / "index_template.html"
+    return path.read_text()
+
+
+class ReportDirMixin:
+    @functools.cached_property
+    def report_dir(self):
+        return Path(self.opt["data_root"]) / "reports" / self.report_subdir
+
+    @functools.cached_property
+    def report_subdir(self):
+        year = parse_cm.paths.parse_load_name(self.source)[-1]
+        return Path(str(year), self.source, f"{self.obsid:05d}")
+
+
+class ObservationFromInfo(ReportDirMixin):
+    def __init__(self, **kwargs):
+        self.__dict__.update(kwargs)
+
+
+@dataclass(repr=False, kw_only=True)
+class Observation(razl.observations.Observation, ReportDirMixin):
+    obs_next: Optional["Observation"] = None
+    obs_prev: Optional["Observation"] = None
+
+    @functools.cached_property
+    def manvr_event(self):
+        """Provide the maneuver event leading to this observation."""
+        raise_sporadic_exc_for_testing()
+        # Manvr leading to this observation. Remember the Manvr class includes the
+        # maneuver and info about the dwell.
+        manvr = self.manvrs[-1]
+        start = manvr.start - 1 * u.min
+        stop = manvr.stop + 5 * u.min
+        manvrs = ke.manvrs.filter(start, stop)
+        if len(manvrs) == 0:
+            # Most commonly because telemetry does not include this maneuver yet.
+            out = None
+        elif len(manvrs) > 2:
+            # Should never have 3 or more manvrs in a row.
```

Comment:
The current guideline is 2 but are you planning to process the mission?

Reply:
That would be the idea. Are there known cases with more than 2 maneuvers in a row?

Reply:
The last one I see in kadi manvrs is 2 no-catalog pointings in https://icxc.cfa.harvard.edu/mp/mplogs/2019/DEC0219/oflsa/starcheck.html